### PR TITLE
Use react-freeze to optimize TabController pages

### PR DESCRIPTION
**Pull request by ethanshar**
## Description
Use [react-freeze](https://github.com/software-mansion-labs/react-freeze) to optimize functionality of TabController.TabPage so pages that are not focused will not trigger renders

## Changelog
Use react-freeze to optimize functionality of TabController.TabPage so pages that are not focused will not trigger renders

### CODE REVIEW

File: src/components/tabController/TabPage.tsx [hunk 1]
```diff
@@ -49,9 +51,14 @@ export default function TabPage({
   useAnimatedReaction(() => {
     return targetPage.value === index;
   },
-  isActive => {
+  (isActive, wasActive) => {
```
Comment (M-i-k-e-l) [inline src/components/tabController/TabPage.tsx:54]: I think `wasActive` is actually `previousActive` or `previousActiveState`

File: src/components/tabController/TabPage.tsx [hunk 2]
```diff
@@ -49,9 +51,14 @@ export default function TabPage({
...
     if (isActive) {
       runOnJS(lazyLoad)();
+      runOnJS(setFocused)(true);
```
Comment (M-i-k-e-l) [inline src/components/tabController/TabPage.tsx:57]: Do we want to make sure this runs only once `lazyLoad` has finished?

File: src/components/tabController/TabPage.tsx [hunk 3]
```diff
@@ -49,9 +51,14 @@ export default function TabPage({
...
+    }
+
+    if (wasActive && wasActive !== isActive) {
```
Comment (M-i-k-e-l) [inline src/components/tabController/TabPage.tsx:60]: Why not `wasActive && !isActive ` or (to be sure we don't have a race condition):
```
if (isActive) {
  ...
} else if (wasActive) {
  runOnJS(setFocused)(false);
}
```

File: demo/src/screens/componentScreens/TabControllerScreen/tab3.tsx [hunk 4]
```diff
@@ -13,7 +14,16 @@ class Tab2 extends Component {
...
+    /* Uncomment to test TabPage freeze functionality when the page lose focus */
+    // setInterval(() => {
+    //   this.setState({counter: this.state.counter + 1});
+    // }, 1000);
+  }
+
+  componentDidUpdate() {
+    if (this.state.counter % 3 === 0) {
+      console.warn('freeze counter', this.state.counter);
+    }
```
Comment (M-i-k-e-l) [inline demo/src/screens/componentScreens/TabControllerScreen/tab3.tsx:26]: Not sure about the logic here, but either way: did you leave it uncommented on purpose?

File: src/components/tabController/TabPage.tsx (hunk 1 above)
Comment (ethanshar) [reply src/components/tabController/TabPage.tsx:54]: Not sure I understand? 
You mean the name is not good? 
what's the different between wasActive and previousActive?

File: src/components/tabController/TabPage.tsx (hunk 2 above)
Comment (ethanshar) [reply src/components/tabController/TabPage.tsx:57]: Don't think it's necessary 
They don't have to be coupled

File: src/components/tabController/TabPage.tsx (hunk 3 above)
Comment (ethanshar) [reply src/components/tabController/TabPage.tsx]: I change it.. now it's better. 
Apparently we got to this part of code more than once, after adding `currentPage` to dependency array of `useAnimatedReaction` it fixed it.. so we can rely completely on `wasActive`

File: demo/src/screens/componentScreens/TabControllerScreen/tab3.tsx (hunk 4 above)
Comment (ethanshar) [reply demo/src/screens/componentScreens/TabControllerScreen/tab3.tsx:26]: Yes, but I can comment it as well

Comment (ethanshar) [reply demo/src/screens/componentScreens/TabControllerScreen/tab3.tsx]: done

**Comment by ethanshar** [comment]
@M-i-k-e-l 
I fix the logic a little bit.. 
I fixed how we handle pages that are near the active page - this is relevant when using PageCarousel

File: src/components/tabController/TabPage.tsx (hunk 1 above)
Comment (M-i-k-e-l) [reply src/components/tabController/TabPage.tsx:54]: It was a little confusing to me, not sure why. I think it's more standard to call it `prev` or `previous`.
The API for `useAnimatedReaction` is:
```
export function useAnimatedReaction<D>(
  prepare: () => D,
  react: (prepareResult: D, preparePreviousResult: D | null) => void,
  deps?: DependencyList
): void;
```

This is also the case in other methods, but if you think it's more clear then I'm fine with it.

File: src/components/tabController/TabPage.tsx [hunk 5]
```diff
@@ -49,18 +49,29 @@ export default function TabPage({
...
+    if (wasActive) {
       runOnJS(setFocused)(false);
     }
-  });
+  },
+  [currentPage]);
+
+  /* Handle freeze for pages that near the active page (relevant for carousel page) */
+  useAnimatedReaction(() => {
+    return asCarousel && currentPage.value - 1 === index || currentPage.value + 1 === index;
```
Comment (M-i-k-e-l) [inline src/components/tabController/TabPage.tsx:67]: I think we're missing parentheses:
```
return asCarousel && (currentPage.value - 1 === index || currentPage.value + 1 === index);
```

**Review by M-i-k-e-l — COMMENTED** [review]
As we saw, adding the following log reveals problems in the logic of the frozen tabs:
```
console.log('Ethan', index, !shouldLoad || !focused);
```

File: src/components/tabController/TabPage.tsx (hunk 3 above)
Comment (ethanshar) [reply src/components/tabController/TabPage.tsx]: @M-i-k-e-l 
I changed the implementation again, I think it's even simpler now
It has a single `useAnimatedReaction` so I think it's more bug prune 
Let me know if you find something 😬

File: src/components/tabController/TabPage.tsx [hunk 6]
```diff
@@ -47,13 +49,25 @@ export default function TabPage({
...
   useAnimatedReaction(() => {
-    return targetPage.value === index;
+    return currentPage.value;
   },
-  isActive => {
+  (currentPage, previousPage) => {
+    const isActive = currentPage === index;
+    const wasActive = previousPage === index;
+    const nearActive = asCarousel && (currentPage - 1 === index || currentPage + 1 === index);
+    const wasNearActive = asCarousel && previousPage && (previousPage - 1 === index || previousPage + 1 === index);
```
Comment (M-i-k-e-l) [inline src/components/tabController/TabPage.tsx:58]: I think `previousPage` should be `previousPage !== null` (for when it's `0`).

**Review by M-i-k-e-l — COMMENTED** [review]
When on the Posts tab (index=1), do we want the Reviews (index=2) to load? It does not.

File: src/components/tabController/TabPage.tsx (hunk 5 above)
Comment (ethanshar) [reply src/components/tabController/TabPage.tsx]: Fixed that

File: src/components/tabController/TabPage.tsx (hunk 6 above)
Comment (ethanshar) [reply src/components/tabController/TabPage.tsx]: Right! 
Fixed

**Comment by ethanshar** [comment]
> As we saw, adding the following log reveals problems in the logic of the frozen tabs:
> 
> ```
> console.log('Ethan', index, !shouldLoad || !focused);
> ```

The right thing to check is `focused` and not both IMO 
Cause you want to check the focus logic (which seems like it's working) 
The freeze also affected by shouldLoad which might give different results

File: src/components/tabController/TabPage.tsx (hunk 5 above)
Comment (M-i-k-e-l) [reply src/components/tabController/TabPage.tsx]: I think I forgot to delete this comment when I sent the review

**Comment by M-i-k-e-l** [comment]
> > As we saw, adding the following log reveals problems in the logic of the frozen tabs:
> > ```
> > console.log('Ethan', index, !shouldLoad || !focused);
> > ```
> 
> The right thing to check is `focused` and not both IMO Cause you want to check the focus logic (which seems like it's working) The freeze also affected by shouldLoad which might give different results

But if we freeze when `!shouldLoad`, then why have a different logic for the carousel? Assuming it has the same logic for `shouldLoad`, it'll be frozen or not at the same time, wouldn't it?

**Review by M-i-k-e-l — APPROVED** [review]

**Pull request merged**